src/atomvm_esp32_flash_provider.erl

```diff
@@ -185,3 +271,80 @@ get_avm_file(App) ->
+    rebar_api:info("Reading application partition offset from device...", []),
+    try esp_part_dump:read_app_offset(Esptool, Port, PartName, TempFile) of
+        Offset ->
+            Offset
+    catch
+        error:invalid_partition_table ->
+            rebar_api:abort("Invalid partition data!", []);
+        error:no_device ->
+            rebar_api:abort("No ESP32 device attached!", []);
+        error:no_device_dump ->
+            rebar_api:abort("No partition table retrieved from device.", []);
+        error:{partition_not_found, Partition} ->
+            rebar_api:error("The partition ~s was not found in device partition table!", [Partition]),
+            rebar_api:abort(
+                "When using a custom partition table always specify the 'app_partition' NAME.", []
+            );
+        error:{invalid_subtype, Type} ->
+            rebar_api:abort("The partition ~s was found, but used invalid subtype ~w.", [
+                PartName, Type
+            ]);
+        error:invalid_partition_type ->
+            rebar_api:abort(
+                "The partition ~s is not a data partition!~nOnly data partitions may be used for AtomVM applications",
+                [
+                    PartName
+                ]
+            );
+        error:corrupt_partition_data ->
+            rebar_api:abort("Fatal error! Corrupt partition table data!", []);
+        error:no_data_partitions ->
+            rebar_api:abort("The partition table on the device contains no data partitions");
+        error:Error ->
+            rebar_api:abort("Unexpected error reading partition table from device, ~p.", [Error])
+    end.
+
+validate_offset(Esptool, Port, Offset, Partition, TempFile) ->
+    rebar_api:debug("Checking that offset aligns with partition from table on device...", []),
+    try esp_part_dump:partition_at_offset(Esptool, Port, Offset, TempFile) of
+        {error, {Name, Reason}} ->
+            rebar_api:abort(
+                "The configured offset 0x~.16B contains partition ~s, which cannot be used for reason ~p.~n",
+                [Offset, Name, Reason]
+            );
+        {Found, _Type} ->
+            rebar_api:warn(
+                "The discovered partition ~s at offset 0x~.16B will be used, not the expected partition named ~s.",
+                [Found, Offset, Partition]
+            ),
+            {ok, Offset}
+    catch
+        _:no_device ->
+            rebar_api:abort("No ESP32 device attached!", []);
+        _:corrupt_partition_data ->
+            rebar_api:abort("Fatal error! Corrupt partition table data!", []);
+        C:E:S ->
+            rebar_api:debug("Class=~p, Error=~p~nSTACKTRACE:~n~p~n", [C, E, S]),
+            rebar_api:abort(
+                "Unexpected error reading partition table from device. Error = ~p.",
+                [E]
+            )
+    end.
+
+%% @private
+get_part_tempfile(State) ->
+    OutDir = filename:absname(rebar_dir:base_dir(State)),
```

Comment:
We probably should call mktemp(1) here

Reply:
Indeed. When I was first implementing this I intended to look into OTP methods for using temp files, but by the time I got the tricky bits working I forget to go back and address this.